# Prueba final (20 %)

- **Caso:** La red aérea centroamericana
- **Curso:** BCD5105 Modelado matemático
- **Profesor:** Jordy Alfaro Brenes
- **Modalidad:** Grupal, sincrónica vía Zoom
- **Duración:** 90 minutos
- **Materiales permitidos:** Google Colab con <code>notebooks/PruebaFinal_celdas.ipynb</code>, calculadora y los archivos <code>datos/aeropuertos.csv</code> y <code>datos/rutas.csv</code>.
- **Indicación original:** No se permitía inteligencia artificial generativa durante la aplicación de la prueba.
- **Valor:** 100 puntos.
  - **Parte A:** ejecución y análisis en Colab, 60 puntos.
  - **Parte B:** interpretación de resultados y conceptos, 40 puntos.
- **Integrantes:**
  - Siloé Campos
  - Jason Courrau
  - Gabriel Corrales
  - David Mora

---

## El caso: equipo de planificación de una aerolínea regional

Una aerolínea regional opera en Centroamérica. El equipo de ciencia de datos recibió <code>aeropuertos.csv</code>, con los 10 aeropuertos internacionales principales de la región y sus coordenadas, y <code>rutas.csv</code>, con las 29 rutas comerciales registradas entre ellos y sus distancias de gran círculo en kilómetros. Toda la prueba utiliza estos datos, mientras que las tarifas, costos y probabilidades adicionales se consideran supuestos del enunciado.

# Parte A: ejecución y análisis en Colab (60 puntos)

## A1. La ruta de Liberia a Roatán (15 puntos)

La aerolínea quiere llevar carga de Liberia (LIR) a Roatán (RTB) usando únicamente vuelos presentes en <code>rutas.csv</code>. La celda A1 construye la red de 10 aeropuertos y 29 rutas y ejecuta Dijkstra.

<pre>Distancias mínimas desde LIR (km):
LIR:   0
SJO: 161
MGA: 482
SAL: 496
PTY: 696
GUA: 699
TGU: 706
SAP: 751
RTB: 916
BZE: 959

Ruta más corta LIR → RTB: LIR – SAL – RTB (916 km, 1 escala)</pre>

### A1.a (5 puntos)
**Un analista propone LIR–SAL–SAP–RTB. Calculen su longitud y expliquen por qué Dijkstra no la eligió.**

La propuesta suma los tramos LIR–SAL (496 km), SAL–SAP (255 km) y SAP–RTB (178 km), con lo que alcanza el siguiente total.

$$496 + 255 + 178 = 929\text{ km}$$

| Ruta | Distancia | Escalas |
|---|---:|---:|
| LIR–SAL–SAP–RTB | 929 km | 2 |
| LIR–SAL–RTB | 916 km | 1 |

La salida muestra que SAL queda a 496 km, SAP a 751 km y RTB a 916 km. Desde SAL, Dijkstra alcanza RTB con 916 km, por lo que continuar primero a SAP produciría 929 km. Por ello se descarta la propuesta, que es 13 km más larga.

### A1(b) (4 puntos)

**Expliquen por qué la lista de distancias aparece ordenada de menor a mayor.**

Dijkstra selecciona en cada iteración el nodo abierto con menor distancia acumulada y lo cierra, dejando definitiva su distancia porque todas las aristas tienen pesos no negativos. Por eso los nodos se cierran en orden no decreciente de distancia desde el origen. 

La celda presenta las distancias finales en ese mismo orden por lo que si existieran empates, varios nodos podrían compartir la misma distancia.

### A1(c) (6 puntos)

**Cambien el origen y destino a SJO y RTB, luego reporten la ruta, distancia y escalas, y expliquen por qué la conexión elegida es mejor.**

La ruta mínima es **SJO–TGU–RTB**, con una escala en Tegucigalpa. Su distancia total se obtiene de la siguiente manera.

$$558 + 262 = 820\text{ km}$$

| Conexión | Tramos | Distancia total |
|---|---:|---:|
| SJO–TGU–RTB | 558 + 262 | 820 km |
| SJO–SAP–RTB | 728 + 178 | 906 km |
| SJO–SAL–RTB | 652 + 420 | 1,072 km |
| SJO–GUA–RTB | 855 + 470 | 1,325 km |

Como no existe vuelo directo SJO–RTB, TGU es el mejor nodo de conexión porque genera el menor kilometraje total entre las alternativas de una escala.

## A2. Repartir la flota entre Panamá y Guatemala (15 puntos)

Cada vuelo redondo SJO–PTY deja USD 500 y consume 3 horas, mientras que cada vuelo SJO–GUA deja USD 600 y consume 4 horas. Hay 60 horas semanales de flota y 18 salidas.

<pre>Plan óptimo: 12 vuelos a Panamá y 6 a Guatemala por semana
Ganancia semanal máxima: 9,600 USD
Horas de flota usadas: 60 de 60
Salidas usadas: 18 de 18
Precio sombra de una hora adicional: 100 USD
Precio sombra de un slot adicional: 200 USD</pre>

### A2(a) (4 puntos)

**¿Qué significa que ambas restricciones estén activas y cómo se relaciona con los precios sombra positivos?**

| Recurso | Uso | Límite | Recursos sin usar | Precio sombra |
|---|---:|---:|---:|---:|
| Horas de flota | 60 | 60 | 0 | USD 100 por hora |
| Slots de salida | 18 | 18 | 0 | USD 200 por slot |

Ambas restricciones son vinculantes porque se utiliza toda su disponibilidad. En esta solución, cada recurso limita la ganancia y una unidad adicional permite aumentarla marginalmente, de ahí los precios sombra positivos. Si una restricción tiene recursos disponibles sin utilizar, su precio sombra normalmente es cero. Sin embargo, que una restricción esté activa no significa necesariamente que una unidad adicional del recurso aumentará la ganancia.

### A2(b) (5 puntos)

**Si una hora adicional y un slot adicional tienen el mismo costo, ¿cuál debería negociarse y qué mide el precio sombra?**

| Alternativa | Aumento marginal de la ganancia |
|---|---:|
| Una hora adicional | USD 100 |
| Un slot adicional | USD 200 |

Conviene negociar el slot porque aporta el doble de ganancia marginal al mismo costo. El precio sombra mide cuánto cambia la función objetivo al aumentar en una unidad la disponibilidad de un recurso, manteniendo lo demás constante y dentro del rango de sensibilidad en el que no cambia la estructura de la solución óptima.

### A2(c) (6 puntos)

**Cambien HORAS de 60 a 70, luego reporten el nuevo plan y expliquen por qué se desplaza hacia Guatemala.**

| Horas disponibles | Panamá | Guatemala | Ganancia | Horas usadas | Slots usados |
|---:|---:|---:|---:|---:|---:|
| 60 | 12 | 6 | USD 9,600 | 60/60 | 18/18 |
| 70 | 2 | 16 | USD 10,600 | 70/70 | 18/18 |

Las 10 horas adicionales relajan la escasez relativa de flota y permiten sustituir vuelos a Panamá por vuelos a Guatemala. Como los 18 slots siguen limitando el número total de salidas, conviene asignar más a Guatemala, que genera $600 por salida frente a $500 de Panamá. En el nuevo óptimo ambas restricciones continúan vinculantes porque se usan 70 de 70 horas y 18 de 18 slots.

## A3. ¿Abrir la ruta San José–Belice? (15 puntos)

Abrir SJO–BZE produce, en miles de USD, +180 con probabilidad 0.40, +40 con 0.35 y −120 con 0.25. No abrir produce 0 con certeza.

<pre>Utilidad promedio si abre: +55.0 mil USD
Utilidad si no abre: +0.0 mil USD
Años con pérdida: 25.2 %
Peor resultado: −120 mil USD
Utilidad promedio con estudio previo: +85.2 mil USD
Valor de la información: 30.3 mil USD</pre>

### A3(a) (4 puntos)

**¿Qué decisor seguiría el promedio y cuál se fijaría en el peor caso?**

| Criterio | Resultado relevante | Decisión |
|---|---:|---|
| Valor esperado | +55.0 mil USD frente a 0 | Abrir |
| Maximin o peor caso | −120 mil USD frente a 0 | No abrir |

El decisor guiado por valor esperado acepta el riesgo porque la utilidad promedio es positiva. El decisor maximin (criterio de peor caso), más conservador y adverso al riesgo, compara los peores resultados y elige no abrir para evitar la pérdida de 120 mil USD.

### A3(b) (5 puntos)

**Expliquen de dónde sale el valor de 30.3 mil USD del estudio de mercado.**

Sin el estudio, la aerolínea debe decidir antes de conocer la demanda. Como la utilidad promedio de abrir es positiva, abre la ruta y acepta el riesgo de perder 120 mil USD si la demanda es baja.

Con el estudio previo, conoce el escenario antes de decidir. Abre cuando la demanda es alta o media y no abre cuando es baja, por lo que obtiene 0 en vez de perder 120 mil USD. Esto aumenta la utilidad promedio de 55.0 a 85.2 mil USD.

$$
85.2-55.0\approx30.3\text{ mil USD}
$$

La diferencia representa la ganancia promedio adicional obtenida al adaptar la decisión al nivel de demanda.

### A3(c) (6 puntos)

**Cambien las probabilidades a [0.20, 0.30, 0.50], luego reporten el nuevo promedio, la recomendación y el valor de la información.**

| Resultado simulado | Escenario original | Escenario pesimista |
|---|---:|---:|
| Utilidad promedio al abrir | +55.0 mil USD | −13.9 mil USD |
| Años con pérdida | 25.2 % | 50.6 % |
| Recomendación por valor esperado | Abrir | No abrir |
| Utilidad con estudio | +85.2 mil USD | +46.8 mil USD |
| Valor impreso de la información | 30.3 mil USD | 60.8 mil USD |

La recomendación cambia porque abrir pasa a tener utilidad promedio negativa. Según la formulación suministrada, el estudio gana valor al identificar con mayor frecuencia la demanda baja y evitar la pérdida de 120 mil USD.

> **Observación técnica** El valor de 60.8 mil USD compara la decisión con estudio frente a abrir siempre. Si se comparara con la mejor decisión sin estudio, que es no abrir y obtener 0, el valor simulado sería 46.8 mil USD. Se mantiene 60.8 mil USD porque corresponde al resultado oficial proporcionado por el profesor.

## A4. ¿Cuál promoción mostrar? (15 puntos)

El registro histórico contiene 58 compras en 200 vistas para A, 18 en 80 para B y 5 en 20 para C. Las tasas verdaderas de la simulación son A = 0.25, B = 0.18 y C = 0.40.

<pre>EL TERCO: 1,939 compras en 5,000 visitantes
Visitantes que vieron C: 91.9 %
Reparto: A 4 %, B 4 %, C 92 %

EL APOSTADOR: 1,978 compras en 5,000 visitantes
Visitantes que vieron C: 97.6 %
Reparto: A 1 %, B 1 %, C 98 %</pre>

### A4(a) (4 puntos)

**¿Por qué el registro histórico hace parecer mejor a A y qué ocurriría si se eligiera para siempre?**

| Promoción | Compras/vistas | Tasa observada | Error estándar | Intervalo aproximado del 95 % |
|---|---:|---:|---:|---:|
| A | 58/200 | 0.290 | 0.032 | [0.23, 0.35] |
| C | 5/20 | 0.250 | 0.097 | [0.06, 0.44] |

La estimación de C es mucho menos precisa porque usa apenas 20 vistas, lo que significa que su tasa observada puede quedar por debajo de A por variación muestral aunque su tasa verdadera sea 0.40. Elegir A para siempre eliminaría la exploración, impediría acumular evidencia sobre C y mantendría una pérdida esperada de aproximadamente 0.15 compras por visita respecto a la mejor promoción.

### A4(b) (5 puntos)

**¿Por qué el terco llega a 91.9 % de C y el apostador a 97.6 %?**

| Política | Exploración | Tráfico hacia C | Efecto estructural |
|---|---|---:|---|
| Terco | 10 % al azar permanentemente | 91.9 % | Mantiene un costo fijo de exploración |
| Apostador | Disminuye conforme aprende | 97.6 % | Concentra gradualmente el tráfico en C |

Ambas políticas descubren que C es la mejor, pero el terco nunca reduce su exploración aleatoria. El apostador actualiza su incertidumbre y explora cada vez menos, por lo que dedica una proporción mayor de visitantes a C.

### A4(c) (6 puntos)

**Cambien VISITANTES de 5,000 a 50,000, luego reporten las compras y expliquen qué ocurre con la desventaja del terco.**

| Política | 5,000 visitantes | 50,000 visitantes |
|---|---:|---:|
| Terco | 1,939 compras y 91.9 % vio C | 19,374 compras y 93.1 % vio C |
| Apostador | 1,978 compras y 97.6 % vio C | 19,948 compras y 99.7 % vio C |
| Diferencia | 39 compras | 574 compras |

La desventaja absoluta crece de 39 a 574 compras. El terco mantiene una fracción permanente de exploración y acumula pérdidas, mientras el apostador concentra casi todo el tráfico en C, por lo que “el terco nunca deja de perder”.

# Parte B: análisis conceptual y de código (40 puntos)

## B1. El modelo de tiempos de vuelo (14 puntos)

El modelo <code>t = θd</code> relaciona el tiempo de vuelo en minutos con la distancia en kilómetros. Se probaron dos tasas de aprendizaje mediante descenso de gradiente.

In [ ]:
import numpy as np

d = np.array([234, 463, 236, 544, 1358, 470, 203, 296, 855, 361, 696, 496,
              161, 816, 345, 321, 1161, 1168, 539, 1018, 420, 178, 262, 255,
              652, 210, 728, 172, 558], dtype=float)
t = np.round(d / 800 * 60)  # minutos de vuelo observados

def gradiente(theta):
    """Calcula el gradiente de la función de error"""
    return -2 * np.mean(d * (t - theta * d))  # derivada del error cuadrático medio

for lr, nombre in [(3e-7, "tasa A = 0.0000003"), (4e-6, "tasa B = 0.000004")]:
    theta = 0.0
    print(f"--- {nombre} ---")
    print(f"{'iteración':>10}{'theta':>12}{'error medio':>15}")
    for k in range(1, 41):
        theta = theta - lr * gradiente(theta)
        if k in (1, 2, 3, 5, 10, 20, 40):
            err = np.mean((t - theta * d) ** 2)
            print(f"{k:>10}{theta:>14.5g}{err:>15.4g}")

--- tasa A = 0.0000003 ---
 iteración       theta    error medio
         1      0.017141           1276
         2      0.030366          759.7
         3       0.04057          452.3
         5      0.054517          160.3
        10      0.069424          12.05
        20      0.074615         0.1354
        40      0.075032        0.06839
--- tasa B = 0.000004 ---
 iteración       theta    error medio
         1       0.22854           8972
         2      -0.23902      3.755e+04
         3       0.71753      1.572e+05
         5        2.7642      2.753e+06
        10       -96.299      3.536e+09
        20   -1.2378e+05      5.834e+15
        40   -2.0421e+11      1.588e+28


### B1(a) (4 puntos)

**Interpreten θ ≈ 0.0750 y calculen la velocidad de crucero.**

El parámetro representa 0.075 minutos por kilómetro. Al invertirlo y convertir minutos a horas se obtiene el siguiente resultado.

$$v = \frac{1}{0.075}\frac{\text{km}}{\text{min}}\times 60\frac{\text{min}}{\text{h}} = 800\text{ km/h}$$

Por ejemplo, para GUA–PTY se obtiene <code>1,358 × 0.075 = 101.85 minutos</code>, que redondea a los 102 minutos de <code>rutas.csv</code>.

### B1(b) (5 puntos)

**Diagnostiquen el comportamiento de ambas tasas usando números de la salida.**

| Tasa | Comportamiento de θ | Error medio | Diagnóstico |
|---|---|---|---|
| A = 0.0000003 | 0.017141 → 0.075032 | 1,276 → 0.06839 | Converge de forma estable |
| B = 0.000004 | 0.22854 → −0.23902 → 0.71753 → −2.0421×10¹¹ | 8,972 → 1.588×10²⁸ | Diverge por pasos excesivos |

La tasa A reduce el error y estabiliza θ cerca del mínimo. La tasa B sobrepasa repetidamente el mínimo, alterna el signo de θ y aumenta el error. El valor negativo de la segunda iteración implicaría tiempo negativo para una distancia positiva, lo cual carece de sentido físico.

### B1(c) (5 puntos)

**Den dos razones para usar descenso de gradiente y una precaución aplicable a redes neuronales.**

| Razón | Aplicación práctica |
|---|---|
| Escalabilidad | Evita resolver directamente sistemas muy grandes y admite variantes por lotes o minilotes. |
| Generalidad | Optimiza modelos complejos y no lineales sin solución cerrada, como las redes neuronales. |

En mínimos cuadrados no es necesario calcular explícitamente una inversa porque pueden usarse factorizaciones o sistemas lineales, aunque su costo puede ser alto en gran escala. La precaución general es controlar la tasa de aprendizaje. Si es demasiado grande, los pesos oscilan, la pérdida diverge y el entrenamiento puede fallar.

## B2. La gira de inspección (14 puntos)

Una avioneta sale de SJO, visita una vez los otros cinco aeropuertos de la subred y regresa a SJO. Las distancias son de gran círculo.

<pre>FUERZA BRUTA
120 giras evaluadas
SJO – LIR – SAL – SAP – RTB – TGU – SJO
1909.2 km

RECOCIDO SIMULADO
400 pasos
SJO – TGU – RTB – SAP – SAL – LIR – SJO
1909.2 km</pre>

### B2(a) (5 puntos)

**¿Los métodos encontraron soluciones distintas o la misma?**

| Método | Gira |
|---|---|
| Fuerza bruta | SJO → LIR → SAL → SAP → RTB → TGU → SJO |
| Recocido simulado | SJO → TGU → RTB → SAP → SAL → LIR → SJO |

Es la misma gira recorrida en sentidos opuestos. Como las distancias de gran círculo son simétricas, invertir el sentido conserva cada tramo y el total de 1,909.2 km.

### B2(b) (5 puntos)

**¿Qué papel juega la temperatura y por qué se aceptan a veces giras peores?**

| Estado | Comportamiento |
|---|---|
| Temperatura alta | Acepta con mayor probabilidad movimientos peores y explora regiones distintas. |
| Temperatura baja | Reduce gradualmente esa probabilidad y concentra la búsqueda en mejoras. |
| Sin aceptar deterioros | Se comportaría como búsqueda local y podría quedar atrapado en un mínimo local. |

Aceptar temporalmente una gira más larga permite escapar de soluciones localmente buenas pero globalmente subóptimas. Incluso a baja temperatura puede aceptarse un deterioro, aunque con probabilidad cada vez menor. Solo en el límite de temperatura cero se aceptarían únicamente mejoras.

### B2(c) (4 puntos)

**¿Cuándo se justifica una heurística y qué garantía se sacrifica?**

Con 6 aeropuertos, la fuerza bruta evalúa 120 giras y resulta más económica que los 400 pasos usados por el recocido en esta ejecución.

Con 15 aeropuertos, evaluar 14! ≈ 8.7×10¹⁰ giras vuelve inviable la búsqueda exhaustiva y justifica una heurística que produzca una buena solución en tiempo razonable.

A cambio de esa escalabilidad se sacrifica la garantía de encontrar el óptimo global. Que 400 pasos hayan coincidido con el óptimo en el caso pequeño no asegura el mismo resultado en una red mayor.

## B3. Preguntas conceptuales (12 puntos)

### B3(i) (4 puntos)

**SAL tiene nueve conexiones. ¿Basta para considerarlo el nodo más importante? Propongan otra medida.**

No. El grado mide importancia local mediante el número de conexiones directas, pero no describe por sí solo el papel global del nodo. La centralidad de intermediación mide con qué frecuencia un aeropuerto aparece en los caminos más cortos entre otros pares y, por tanto, su función como puente. Esta medida podría ordenar los aeropuertos de manera diferente. Para afirmar cuál es el más importante bajo ese criterio habría que calcularla sobre la red, no inferirla únicamente por ubicación geográfica.

### B3(ii) (4 puntos)

**Mencionen una conclusión que podría cambiar con datos actuales y expliquen por qué un modelo hereda las limitaciones de sus datos.**

La ruta mínima LIR–SAL–RTB podría cambiar si actualmente existe una conexión nueva o dejó de operar alguna ruta registrada en 2014. El modelo solo representa los nodos, conexiones y pesos presentes en su fuente histórica. Por ello, cualquier omisión, desactualización o sesgo de los datos se transmite a sus resultados y limita su validez para describir la red actual.

### B3(iii) (4 puntos)

**Usen la ecuación de Bellman para explicar por qué la mayor ganancia inmediata puede no ser óptima.**

Bellman expresa el valor como la recompensa inmediata más el valor descontado del estado futuro mediante la relación <code>V(s) = r + γV(s')</code>. Saturar hoy la ruta más rentable podría maximizar el ingreso inmediato, pero dejar a la aerolínea sin capacidad para atender otras rutas, responder a cambios de demanda o realizar mantenimiento. Una asignación más equilibrada puede sacrificar parte de la ganancia actual y, al mismo tiempo, conducir a un estado futuro más flexible y valioso, maximizando el retorno total de largo plazo.